jsapp/js/components/modalForms/formMedia.es6

```diff
@@ -0,0 +1,277 @@
+   * setup
+   */
+
+  componentDidMount() {
+    actions.media.loadMedia.completed.listen(this.onGetMediaCompleted);
+    actions.media.uploadMedia.completed.listen(this.onUploadCompleted);
+    actions.media.uploadMedia.failed.listen(this.onUploadFailed);
+    actions.media.deleteMedia.completed.listen(this.onGetMediaCompleted);
+    actions.media.deleteMedia.failed.listen(this.onDeleteMediaFailed);
+
+    actions.media.loadMedia(this.props.asset.uid);
+  }
+
+  /*
+   * action listeners
+   */
+
+  onGetMediaCompleted(uploadedAssets) {
+    this.setState({
+      uploadedAssets: uploadedAssets.results,
+      isUploadFilePending: false,
+      isUploadURLPending: false,
+      isVirgin: false,
+    });
+  }
+
+  onUploadFailed(response) {
+    this.setState({
+      fieldsErrors: response.responseJSON,
+      isUploadFilePending: false,
+      isUploadURLPending: false
+    });
+  }
+
+  /*
+   * Utilities
+   */
+
+  toBase64(file) {
+    return new Promise(function(resolve, reject) {
+      var reader = new FileReader();
+      reader.readAsDataURL(file);
+      reader.onload = function() {
+          return resolve(reader.result);
+      };
+      reader.onerror = function(error) {
+          return reject(error);
+      };
+    });
+  }
+
+  uploadMedia(formMediaJSON) {
+    // Reset error message before uploading again
+    this.setState({fieldsErrors: {}});
+    actions.media.uploadMedia(this.props.asset.uid, formMediaJSON);
+  }
+
+  /*
+   * DOM listeners
+   */
+
+  onFileDrop(files) {
+    if (files.length >= 1) {
+      this.setState({isUploadFilePending: true});
+
+      files.forEach(async (file) => {
+        var base64File = await this.toBase64(file);
+	    var formMediaJSON = {
+	      description: 'default',
+          file_type: 'form_media',
+          metadata: JSON.stringify({filename: file.name}),
+	      base64Encoded: base64File
+	    };
+        this.uploadMedia(formMediaJSON);
+      });
+    }
+  }
+
+  onInputURLChange(inputURL) {
+    this.setState({inputURL: inputURL});
+  }
+
+  onSubmitURL() {
+    var url = this.state.inputURL;
+
+    if (url === '') {
+      alertify.warning(t('URL is empty!'));
+    } else {
+      this.setState({
+        isUploadURLPending: true,
+        inputURL: ''
+      });
+
+      var formMediaJSON = {
+        description: 'default',
+        file_type: 'form_media',
+        metadata: JSON.stringify({redirect_url: url}),
+      };
+      this.uploadMedia(formMediaJSON);
+    }
+  }
+
+  onDeleteMedia(url) {
+    actions.media.deleteMedia(this.props.asset.uid, url);
+  }
+
+  uploadedAssetsIsEmpty() {
+    return (
+      this.state.uploadedAssets === null ||
+      this.state.uploadedAssets === undefined
+    );
+  }
+
+  /*
+   * rendering
+   */
+
+  renderButton() {
+    const buttonClassNames = ['formBuilder-header__button', 'formBuilder-header__button--save'];
+    if (this.state.isUploadURLPending) {
+      buttonClassNames.push('formBuilder-header__button--savepending');
```

Comment:
This causes a knock-on effect for the formbuilder `SAVE` button. If we do this then we would need to change the following in `editableForm.es6`:
```diff
diff --git a/jsapp/js/editorMixins/editableForm.es6 b/jsapp/js/editorMixins/editableForm.es6
index 13140585..71eb1aea 100644
--- a/jsapp/js/editorMixins/editableForm.es6
+++ b/jsapp/js/editorMixins/editableForm.es6
@@ -579,7 +579,7 @@ export default assign({
           <bem.FormBuilderHeader__cell m={'buttonsTopRight'} >
             <bem.FormBuilderHeader__button
               m={['save', {
-                savepending: this.state.asset_updated === update_states.PENDING_UPDATE,
+                'save-pending': this.state.asset_updated === update_states.PENDING_UPDATE,
                 savefailed: this.state.asset_updated === update_states.SAVE_FAILED,
                 saveneeded: this.needsSave(),
               }]}
diff --git a/jsapp/scss/components/_kobo.form-builder.scss b/jsapp/scss/components/_kobo.form-builder.scss
index f7c80370..7a63d9f8 100644
--- a/jsapp/scss/components/_kobo.form-builder.scss
+++ b/jsapp/scss/components/_kobo.form-builder.scss
@@ -293,7 +293,7 @@ $t-form-builder-aside-open: 200ms;
     }
   }

```

unless you know a cleaner way to do this :D 

But this may be a good opportunity to make the "loading icon button" more general and not use formbuilder css classes, so we can use it in other places? Or we can think of a better way to handle buttons and load in general. Let me know what you think!

Reply:
I don't mind quotes around `'save-pending'` :) Form Builder will be redone sometime in future, so the we will make it better :)